File: csharp/rest-api/RestApi.cs
using System;
using System.Linq;
using Newtonsoft.Json;
using System.Collections.Generic;

public class RestApi
{
    private List<User> database;

    public RestApi(string database) => 
        this.database = JsonConvert.DeserializeObject<List<User>>(database);

    public string Get(string url, string payload = null)
    {
        if(url == "/users")
        {
            if(payload == null)
                return JsonConvert.SerializeObject(database.OrderBy(o => o.name));
            var model = JsonConvert.DeserializeObject<GetUsersModel>(payload);
            var users = database.Where(o => model.users.Contains(o.name));
            return JsonConvert.SerializeObject(users.OrderBy(o => o.name));
        }
        else
            throw new InvalidOperationException();
    }

    public string Post(string url, string payload)
    {
        if(url == "/add")
        {
            var model = JsonConvert.DeserializeObject<AddModel>(payload);
            var newUser = new User(model.user);
            database.Add(newUser);
            return JsonConvert.SerializeObject(newUser);
        }
        else if(url == "/iou")
        {
            var model = JsonConvert.DeserializeObject<IOUModel>(payload);
            var lender = database.Single(o => o.name == model.lender);
            var borrower = database.Single(o => o.name == model.borrower);
            LendTo(lender, borrower, model.amount);
            return JsonConvert.SerializeObject(new [] { lender, borrower }.OrderBy(o => o.name));
        }
        else
            throw new InvalidOperationException();
    }

    private void LendTo(User lender, User borrower, double amount)
    {
        var existingDebt = lender.owes.GetValueOrDefault(borrower.name, 0);
        if(existingDebt > amount)
            LendTo(borrower, lender, existingDebt - amount);
        else if (existingDebt > 0)
        {
            lender.owes.Remove(borrower.name);
            borrower.owed_by.Remove(lender.name);
            LendTo(lender, borrower, amount - existingDebt);
        }
        else
        {
            lender.owed_by[borrower.name] = amount;
            borrower.owes[lender.name] = amount;
        }
    }
}

public struct GetUsersModel
{
    public string[] users { get; set; }
}

public struct AddModel
{
    public string user { get; set; }
}

public struct IOUModel
{
    public string lender { get;set; }
    public string borrower { get;set; }
    public double amount { get;set; }
}

public struct User
{
    public string name { get;set; }
    public SortedDictionary<string, double> owes { get;set; }
    public SortedDictionary<string, double> owed_by { get;set; }
    public double balance => owed_by.Sum(o => o.Value) - owes.Sum(o => o.Value);

    public User(string name)
    {
        this.name = name;
        owes = new SortedDictionary<string, double>();
        owed_by = new SortedDictionary<string, double>();
    }
}

File: fsharp/rest-api/RestApi.fs
module RestApi
open Newtonsoft.Json

type Database = {
    users: User []
}
and User = {
    name: string
    owes: Map<string, float>
    owed_by: Map<string, float>
    balance: float
}

type UserRequest = {
    users: string []
}
type AddRequest = {
    user: string
}
type ChangeRequest = {
    lender: string
    borrower: string
    amount: float
}

let parse<'T> = JsonConvert.DeserializeObject<'T>
let json = JsonConvert.SerializeObject
let notFound () = failwith "url not recognised (404)"

let private updateBalance user =
    let newBalance = 
        (user.owed_by |> Map.toList |> List.sumBy snd)
        - (user.owes |> Map.toList |> List.sumBy snd)
    { user with balance = newBalance }

let private applyChange lender borrower amount =
    let currentOwedByLender = Map.tryFind borrower.name lender.owes |> Option.defaultValue 0.
    if currentOwedByLender = 0. then
        let newLender = { lender with owed_by = Map.add borrower.name amount lender.owed_by }
        let newBorrower = { borrower with owes = Map.add lender.name amount borrower.owes }
        updateBalance newLender, updateBalance newBorrower
    elif currentOwedByLender > amount then
        let newLender = { lender with owes = Map.add borrower.name (currentOwedByLender - amount) lender.owes }
        let newBorrower = { borrower with owed_by = Map.add lender.name (currentOwedByLender - amount) borrower.owed_by }
        updateBalance newLender, updateBalance newBorrower
    elif currentOwedByLender = amount then
        let newLender = { lender with owes = Map.remove borrower.name lender.owes }
        let newBorrower = { borrower with owed_by = Map.remove lender.name borrower.owes }
        updateBalance newLender, updateBalance newBorrower
    else
        let newLender = 
            { lender with 
                owed_by = Map.add borrower.name (amount - currentOwedByLender) lender.owed_by
                owes = Map.remove borrower.name lender.owes }
        let newBorrower = 
            { borrower with 
                owes = Map.add lender.name (amount - currentOwedByLender) borrower.owes
                owed_by = Map.remove lender.name lender.owed_by }
        updateBalance newLender, updateBalance newBorrower

let private updateOwed amount otherName user =
    let current = Map.tryFind otherName user.owed_by |> Option.defaultValue 0.
    let newOwed = 
        Map.add otherName (current - amount) user.owed_by
        |> Map.filter (fun _ value -> value > 0.)
    { user with owed_by = newOwed }


type RestApi(database : string) =

    let mutable database = parse database

    member __.Get(url: string) =
        match url with
        | "/users" -> json database
        | _ -> notFound ()

    member __.Get(url: string, payload: string) =
        match url with
        | "/users" ->
            let request = parse<UserRequest> payload
            let filtered: Database = 
                { database with 
                    users = 
                        database.users 
                        |> Array.filter (fun u -> 
                            Array.contains u.name request.users)
                        |> Array.sortBy (fun u -> u.name) }
            json filtered
        | _ -> notFound ()

    member __.Post(url: string, payload: string)  =
        match url with
        | "/add" ->
            let request = parse<AddRequest> payload
            let newUser = {
                name = request.user
                owes = Map.empty
                owed_by = Map.empty
                balance = 0.
            }
            database <- { database with users = Array.append [|newUser|] database.users }
            json newUser
        | "/iou" ->
            let request = parse<ChangeRequest> payload
            let lender = Array.find (fun u -> u.name = request.lender) database.users
            let borrower = Array.find (fun u -> u.name = request.borrower) database.users
            let newLender, newBorrower = applyChange lender borrower request.amount

            let newUsers = 
                database.users 
                |> Array.map (fun u -> 
                    if u.name = lender.name then newLender 
                    elif u.name = borrower.name then newBorrower 
                    else u)
            database <- { database with users = newUsers }

            let result: Database = 
                { users = Array.sortBy (fun u -> u.name) [|newLender;newBorrower|]}
            json result
        | _ -> notFound ()